File: AlgorithmsPackPythonSolutions/Largest Color Value in a Directed Graph.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        nodes = len(colors)
        graph = [[] for _ in range(nodes)]
        incomingEdges = [0] * nodes
        
        for edge in edges:
            from_node, to_node = edge
            graph[from_node].append(to_node)
            incomingEdges[to_node] += 1
        
        queue = deque()
        colorValues = [[0] * 26 for _ in range(nodes)]
        nodesFound = 0
        result = 0
        
        for i in range(nodes):
            if incomingEdges[i] == 0:
                queue.append(i)
        
        while queue:
            cur = queue.popleft()
            color = ord(colors[cur]) - ord('a')
            colorValues[cur][color] += 1
            nodesFound += 1
            result = max(result, colorValues[cur][color])
            
            for neighbor in graph[cur]:
                incomingEdges[neighbor] -= 1
                if incomingEdges[neighbor] == 0:
                    queue.append(neighbor)
                
                for col in range(26):
                    colorValues[neighbor][col] = max(colorValues[neighbor][col], colorValues[cur][col])
        
        if nodesFound != nodes:
            return -1
        
        return result

File: AlgorithmsPackJavaSolutions/Largest Color Value in a Directed Graph.java
class Solution {
    public int largestPathValue(String colors, int[][] edges) {
        int nodes = colors.length();
        List<List<Integer>> graph = new ArrayList<>();
        int[] incomingEdges = new int[nodes];
        
        for (int i = 0; i < nodes; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] edge : edges) {
            int from = edge[0];
            int to = edge[1];
            graph.get(from).add(to);
            incomingEdges[to]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        int[][] colorValues = new int[nodes][26];
        int nodesFound = 0;
        int result = 0;
        
        for (int i = 0; i < nodes; i++) {
            if (incomingEdges[i] == 0) {
                queue.add(i);
            }
        }
        
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            int color = colors.charAt(cur) - 'a';
            colorValues[cur][color]++;
            nodesFound++;
            result = Math.max(result, colorValues[cur][color]);
            
            for (int neighbor : graph.get(cur)) {
                incomingEdges[neighbor]--;
                if (incomingEdges[neighbor] == 0) {
                    queue.add(neighbor);
                }
                
                for (int col = 0; col < 26; col++) {
                    colorValues[neighbor][col] = Math.max(colorValues[neighbor][col], colorValues[cur][col]);
                }
            }
        }
        
        if (nodesFound != nodes) {
            return -1;
        }
        
        return result;
    }
}
